core/src/main/java/io/atomix/core/Atomix.java

```diff
@@ -263,6 +269,7 @@ public boolean isRunning() {
         .thenComposeAsync(v -> metadataService.stop(), context)
         .thenComposeAsync(v -> messagingService.stop(), context)
         .thenRunAsync(() -> {
+          executorService.shutdown();
```

Comment:
graceful executor shutdown?

```java
executorService.shutdown(); // Disable new tasks from being submitted
try {
    // Wait a while for existing tasks to terminate
    if (!executorService.awaitTermination(10, TimeUnit.SECONDS)) {
        logger.info("Timeout elapsed before termination, forcing shutdown");
        List<Runnable> tasksAwaitingExecution = executorService.shutdownNow(); // Cancel currently executing tasks
        logger.info("Tasks awaiting execution after forced shutdown: " + tasksAwaitingExecution.size());
        // Wait a while for tasks to respond to being cancelled
        if (!executorService.awaitTermination(10, TimeUnit.SECONDS)) {
            logger.error("Pool did not terminate");
        }
    }
} catch (InterruptedException ie) {
    logger.error("Exception occurred while shutting down thread pool", ie);
    // (Re-)Cancel if current thread also interrupted
    executorService.shutdownNow();
    // Preserve interrupt status
    Thread.currentThread().interrupt();
}
logger.info("Thread pool shutdown complete");
```

Reply:
Only potential problem I see is when locks are created with a high timeout the JVM might hang on shutdown while waiting for those? Alternatively the executor could be configured to create daemon threads?

Reply:
@kuujo since we already stopped messaging before this couldn't we just invoke shutdownNow?

Reply:
@kuujo poke

Reply:
I don't have the motivation to fix the merge conflicts ATM :-P

Reply:
okay did